core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -249,9 +249,14 @@ public void run() {
           delayedTransport.endBackoff();
           boolean shutdownDelayedTransport = false;
           Runnable runnable = null;
+          // TransportSet as a channel layer class should not call into transport methods while
+          // holding the lock, thus we call hasPendingStreams() outside of the lock.  It will cause
+          // a _benign_ race where the TransportSet may transition to CONNECTING when there is not
```

Comment:
I don't see where the shutdown error would come from because the delayed transport becomes pass-through to the supplier for newStream(), instead of failing new streams after shutdown.

Instead of calling startNewTransport() here, we would instead enter IDLE. But since we call setTransportSupplier with a supplier that calls obtainActiveTransport, then if the race occurs there will just be a call to obtainActiveTransport that will trigger entering CONNECTING.

The only issue is that the stream will wrapped twice with a DelayedStream. But since the race should only happen at most once per stream, we decided that it wasn't a big deal.